### PR TITLE
Implement a harness for priority_queue sift_down/up/either

### PR DESCRIPTION
**Pull request by angelhof**
Implement a first working harness for `sift_down`, `sift_up`, and `sift_either`. For now the priority queue size is bounded by a very small number due to unwinding complicating the proof.

Depends on: #377 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_priority_queue_s_sift_up/Makefile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+###########
+#
+# Runtime:
+# - 300s for MAX_PRIORITY_QUEUE_ITEMS=3 items
+# - 450s for MAX_PRIORITY_QUEUE_ITEMS=5 items 
+
+
+# Note:
+# In order to reach full coverage we have to unwind
+# log(NUMBER_PRIO_QUEUE_ITEMS) but this number is huge so we only
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_priority_queue_s_sift_up/Makefile]: How can a `log` of a machine integer ever be "huge?"

File: source/priority_queue.c [hunk 2]
```diff
@@ -44,15 +48,20 @@ static void s_swap(struct aws_priority_queue *queue, size_t a, size_t b) {
...
  * in heap order */
 static bool s_sift_down(struct aws_priority_queue *queue, size_t root) {
+    AWS_PRECONDITION(aws_priority_queue_is_valid(queue));
+    AWS_PRECONDITION(root < queue->container.length);
+
     bool did_move = false;
 
     size_t len = aws_array_list_length(&queue->container);
 
+    /* Note: There is an unwinding loop assertion here. */
```
Comment (tautschnig) [inline source/priority_queue.c]: I'm not sure what to make of this comment? In particular as it's part of the source code?

Comment (angelhof) [reply source/priority_queue.c]: I forgot to remove that, thanks for noticing.

File: .cbmc-batch/jobs/aws_priority_queue_s_sift_up/Makefile (hunk 1 above)
Comment (angelhof) [reply .cbmc-batch/jobs/aws_priority_queue_s_sift_up/Makefile]: You are right, that is a wrong and outdated comment that slipped through. I will correct it.

**Comment by tautschnig** [comment]
@angelhof This now requires resolving a merge conflict.

File: source/priority_queue.c [hunk 3]
```diff
@@ -159,10 +176,11 @@ void aws_priority_queue_init_static(
     AWS_ZERO_STRUCT(queue->backpointers);
 
     aws_array_list_init_static(&queue->container, heap, item_count, item_size);
+
+    AWS_POSTCONDITION(aws_priority_queue_is_valid(queue));
 }
 
 bool aws_priority_queue_is_valid(const struct aws_priority_queue *const queue) {
-
```
Comment (tautschnig) [inline source/priority_queue.c]: Unrelated change?

Comment (angelhof) [reply source/priority_queue.c]: Should I separate them from this PR? I added some postconditions to two other functions in `priority_queue.c`.

**Comment by angelhof** [comment]
@ColdenCullen @JonathanHenson @justinboswell @graebm could you review and merge this PR?

File: source/priority_queue.c (hunk 3 above)
Comment (feliperodri) [reply source/priority_queue.c]: I'd argue to keep the pre and postconditions along with the proof in the same PR.

File: source/priority_queue.c [hunk 4]
```diff
@@ -145,7 +158,11 @@ int aws_priority_queue_init_dynamic(
     queue->pred = pred;
     AWS_ZERO_STRUCT(queue->backpointers);
 
-    return aws_array_list_init_dynamic(&queue->container, alloc, default_size, item_size);
+    bool ret = aws_array_list_init_dynamic(&queue->container, alloc, default_size, item_size);
```
Comment (bretambrose) [inline source/priority_queue.c]: It's minor but I would prefer ret to be an error int  (and compare against AWS_OP_SUCCESS) rather than converting to bool and implicitly go back and forth.  It's easy to mentally assume a bool is "success" but here it's the opposite.

Comment (angelhof) [reply source/priority_queue.c]: Resolved, could you merge @bretambrose
